hhds/tree_del.hpp

```diff
@@ -0,0 +1,1488 @@
+    Tree_pos get_last_child_l() const { return last_child_l; }
+
+    // Getters for short child pointers
+    Short_delta get_first_child_s_at(short index) const { 
+        return _get_first_child_s(index); 
+    }
+    Short_delta get_last_child_s_at(short index) const {
+        return _get_last_child_s(index); 
+    }
+
+    // Get the last occupied index in the short delta
+    Short_delta get_last_occupied() const { return last_occupied; }
+
+    // Setters
+    void set_parent(Tree_pos p) { parent = p; }
+    void set_next_sibling(Tree_pos ns) { next_sibling = ns; }
+    void set_prev_sibling(Tree_pos ps) { prev_sibling = ps; }
+    void set_first_child_l(Tree_pos fcl) { first_child_l = fcl; }
+    void set_last_child_l(Tree_pos lcl) { last_child_l = lcl; }
+
+    // Setters for short child pointers
+    void set_first_child_s_at(short index, Short_delta fcs) { 
+        _set_first_child_s(index, fcs); 
+    }
+    void set_last_child_s_at(short index, Short_delta lcs) {
+        _set_last_child_s(index, lcs); 
+    }
+
+    // Set the last occupied index in the short delta
+    void set_last_occupied(Short_delta lo) { last_occupied = lo; }
+
+    // Operators
+    constexpr bool operator==(const Tree_pointers& other) const {
+        for (short i = 0; i < 7; i++) {
+            if (_get_first_child_s(i) != other._get_first_child_s(i) ||
+                _get_last_child_s(i) != other._get_last_child_s(i)) {
+                return false;
+            }
+        }
+
+        return  parent == other.parent && 
+                next_sibling == other.next_sibling &&
+                prev_sibling == other.prev_sibling && 
+                first_child_l == other.first_child_l &&
+                last_child_l == other.last_child_l;
+    }
+
+    constexpr bool operator!=(const Tree_pointers& other) const { return !(*this == other); }
+    void invalidate() { parent = INVALID; }
+// :public
+
+}; // Tree_pointers class
+
+template <typename X>
+class tree {
+private:
+    /* The tree pointers and data stored separately */
+    std::vector<Tree_pointers>              pointers_stack;
+    std::vector<std::optional<X>>           data_stack;
+
+    /* Special functions for sanity */
+    bool _check_idx_exists(const Tree_pos &idx) const noexcept {
+        // idx >= 0 not needed for unsigned int
+        return idx < static_cast<Tree_pos>(data_stack.size());
+    }
+    bool _contains_data(const Tree_pos &idx) const noexcept {
+        return (idx < data_stack.size() && data_stack[idx].has_value());
+    }
+
+    /* Function to add an entry to the pointers and data stack (typically for add/append)*/
+    inline Tree_pos _create_space(const X& data) {
+        // if (pointers_stack.size() >= MAX_TREE_SIZE) {
+        //     throw std::out_of_range("_create_space: Tree size exceeded");
+        // } else if (!_check_idx_exists(parent_index)) {
+        //     throw std::out_of_range("_create_space: Parent index out of range");
+        // }
```

Comment:
Exception handling has been removed on purpose after the `-c opt` flag was used in benchmarking. Since the compiler cannot make decent optimizations due to stack unwinding issues. Maybe iassert or other things can be used instead of exceptions.